server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java

```diff
@@ -781,12 +781,14 @@ void allocateUnassigned() {
          */
         ShardRouting[] unassignedShards = unassigned.drain();
         List<ShardRouting> allUnassignedShards = Arrays.stream(unassignedShards).collect(Collectors.toList());
-        List<ShardRouting> localUnassignedShards = allUnassignedShards.stream()
-            .filter(shard -> RoutingPool.LOCAL_ONLY.equals(RoutingPool.getShardPool(shard, allocation)))
+        List<ShardRouting> nonLocalUnassignedShards = allUnassignedShards.stream()
```

Comment:
Nice find @RS146BIJAY! I can also see that we have heavily used java streams in the PR you referred and which were not being used in the original code. Streams are known to be expensive from both CPU and JVM perspective and the impact depends on how extensively we are using them. Since we are talking about large number of assignments or relocations, it may be impactful.

Just sharing a thought if you want to try getting rid of these streams and measure the impact.

Reply:
Remove unnecessary streams from the code. Actually, major bottleneck here was `List.removeAll `call. Only observed 2 seconds improvements by removing unnecessary streams.

Reply:
Thanks @RS146BIJAY for the update! Ya, `removeAll` is far worse in this case but the 2 sec improvement may still be consistent. For, multiple small lists it is generally more impactful against one big list.